File: typescript/16-proyecto/agente_completo.ts
// Pipeline del agente de revisión de código — equivalente lógico al de Python.

// Cómo ejecutar: make ts SCRIPT=typescript/16-proyecto/agente_completo.ts

import Anthropic from "@anthropic-ai/sdk";
import * as fs from "fs";
import * as path from "path";
import * as crypto from "crypto";
import { execSync } from "child_process";
import * as os from "os";

const SYSTEM_PROMPT = `Eres un agente de revisión de código Python.
Analiza el código con cuidado, usa las herramientas disponibles para verificar comportamiento
cuando sea útil, y produce una revisión técnica estructurada.

Cuando tengas suficiente información, emite el resultado final como JSON con este schema exacto:
{
  "hallazgos": [
    {
      "linea": <número o null>,
      "severidad": "<critical|high|medium|low>",
      "tipo": "<bug|estilo|rendimiento|seguridad>",
      "descripcion": "<descripción concisa del hallazgo>",
      "sugerencia": "<cómo corregirlo>"
    }
  ],
  "resumen": "<párrafo de resumen>"
}`;

const HERRAMIENTAS: Anthropic.Tool[] = [
  {
    name: "read_file",
    description: "Lee el contenido de un archivo del proyecto",
    input_schema: {
      type: "object",
      properties: {
        path: {
          type: "string",
          description: "Ruta relativa al directorio del proyecto",
        },
      },
      required: ["path"],
    },
  },
  {
    name: "run_code",
    description:
      "Ejecuta un fragmento de código Python en sandbox y devuelve stdout/stderr",
    input_schema: {
      type: "object",
      properties: {
        code: { type: "string" },
        timeout: { type: "number", default: 10 },
      },
      required: ["code"],
    },
  },
  {
    name: "search_docs",
    description: "Busca en la documentación técnica del equipo",
    input_schema: {
      type: "object",
      properties: { query: { type: "string" } },
      required: ["query"],
    },
  },
  {
    name: "write_report",
    description: "Escribe el informe final de revisión en disco",
    input_schema: {
      type: "object",
      properties: {
        content: { type: "string" },
        filename: { type: "string" },
      },
      required: ["content", "filename"],
    },
  },
];

interface HallazgoEsperado {
  linea?: number | null;
  severidad: "critical" | "high" | "medium" | "low";
  tipo: "bug" | "estilo" | "rendimiento" | "seguridad";
  descripcion: string;
  sugerencia: string;
}

interface RevisionOutput {
  hallazgos: HallazgoEsperado[];
  resumen: string;
  _meta?: { pasos: number; tokens: { input: number; output: number } };
  _cached?: boolean;
}

function ejecutarHerramienta(
  nombre: string,
  params: Record<string, unknown>,
  proyectoDir: string
): string {
  if (nombre === "read_file") {
    const rutaRelativa = params.path as string;
    const rutaAbs = path.resolve(proyectoDir, rutaRelativa);
    if (!rutaAbs.startsWith(path.resolve(proyectoDir))) {
      return "Error: ruta fuera del directorio del proyecto";
    }
    try {
      return fs.readFileSync(rutaAbs, "utf-8");
    } catch {
      return `Error: archivo '${rutaRelativa}' no encontrado`;
    }
  }

  if (nombre === "run_code") {
    const codigo = params.code as string;
    const timeout = ((params.timeout as number) || 10) * 1000;
    const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), "agente-"));
    const script = path.join(tmpdir, "script.py");
    try {
      fs.writeFileSync(script, codigo);
      const resultado = execSync(`python3 ${script}`, {
        timeout,
        cwd: tmpdir,
        env: { PATH: "/usr/local/bin:/usr/bin:/bin", HOME: tmpdir },
        encoding: "utf-8",
      });
      return JSON.stringify({ stdout: resultado || "(vacío)", stderr: "" });
    } catch (err: unknown) {
      const error = err as { stdout?: string; stderr?: string; message?: string };
      if (error.stderr?.includes("Timeout")) {
        return "Error: timeout de ejecución";
      }
      return JSON.stringify({
        stdout: error.stdout || "(vacío)",
        stderr: error.stderr || error.message || "error desconocido",
      });
    } finally {
      fs.rmSync(tmpdir, { recursive: true, force: true });
    }
  }

  if (nombre === "search_docs") {
    return `[Documentación para '${params.query}': ver /docs/ del proyecto]`;
  }

  if (nombre === "write_report") {
    const reportsDir = path.join(proyectoDir, "reports");
    fs.mkdirSync(reportsDir, { recursive: true });
    const ruta = path.join(reportsDir, params.filename as string);
    fs.writeFileSync(ruta, params.content as string);
    return `Informe escrito en ${ruta}`;
  }

  return `Error: herramienta '${nombre}' desconocida`;
}

function extraerJson(texto: string): RevisionOutput {
  for (let i = 0; i < texto.length; i++) {
    if (texto[i] !== "{") continue;
    let depth = 0;
    for (let j = i; j < texto.length; j++) {
      if (texto[j] === "{") depth++;
      else if (texto[j] === "}") {
        depth--;
        if (depth === 0) {
          try {
            return JSON.parse(texto.slice(i, j + 1)) as RevisionOutput;
          } catch {
            break;
          }
        }
      }
    }
  }
  throw new Error(`No se encontró JSON en output: ${texto.slice(0, 300)}`);
}

async function loopReact(
  codigo: string,
  proyectoDir: string
): Promise<RevisionOutput> {
  const cliente = new Anthropic();
  const mensajes: Anthropic.MessageParam[] = [
    {
      role: "user",
      content: `Revisa este código:\n\n\`\`\`python\n${codigo}\n\`\`\``,
    },
  ];

  const MAX_PASOS = 15;
  const tokensTotal = { input: 0, output: 0 };

  for (let paso = 0; paso < MAX_PASOS; paso++) {
    const respuesta = await cliente.messages.create({
      model: process.env["MODEL"] ?? "claude-sonnet-4-6",
      max_tokens: 4096,
      system: SYSTEM_PROMPT,
      tools: HERRAMIENTAS,
      messages: mensajes,
    });

    tokensTotal.input += respuesta.usage.input_tokens;
    tokensTotal.output += respuesta.usage.output_tokens;

    if (respuesta.stop_reason === "end_turn") {
      const bloque = respuesta.content.find((b) => b.type === "text");
      if (!bloque || bloque.type !== "text") throw new Error("Sin output de texto");
      const revision = extraerJson(bloque.text);
      revision._meta = { pasos: paso + 1, tokens: tokensTotal };
      return revision;
    }

    if (respuesta.stop_reason === "tool_use") {
      mensajes.push({ role: "assistant", content: respuesta.content });
      const resultados: Anthropic.ToolResultBlockParam[] = [];

      for (const bloque of respuesta.content) {
        if (bloque.type === "tool_use") {
          const resultado = ejecutarHerramienta(
            bloque.name,
            bloque.input as Record<string, unknown>,
            proyectoDir
          );
          resultados.push({
            type: "tool_result",
            tool_use_id: bloque.id,
            content: resultado,
          });
        }
      }

      mensajes.push({ role: "user", content: resultados });
    }
  }

  throw new Error(`El agente no terminó en ${MAX_PASOS} pasos`);
}

async function revisar(
  rutaArchivo: string,
  proyectoDir: string
): Promise<RevisionOutput> {
  const codigo = fs.readFileSync(path.join(proyectoDir, rutaArchivo), "utf-8");
  const hashCodigo = crypto
    .createHash("sha256")
    .update(`${rutaArchivo}::${codigo}`)
    .digest("hex");

  // Memoria episódica en archivo JSON (SQLite requería dependencia adicional)
  const dbPath = path.join(proyectoDir, "revisiones.json");
  const db: Record<string, { fecha: string; revision: RevisionOutput }> =
    fs.existsSync(dbPath) ? JSON.parse(fs.readFileSync(dbPath, "utf-8")) : {};

  if (db[hashCodigo]) {
    const cached = db[hashCodigo];
    console.error(`[INFO] Revisión cacheada del ${cached.fecha}`);
    return { ...cached.revision, _cached: true };
  }

  const revision = await loopReact(codigo, proyectoDir);

  db[hashCodigo] = { fecha: new Date().toISOString(), revision };
  fs.writeFileSync(dbPath, JSON.stringify(db, null, 2));

  return revision;
}

// Punto de entrada
const args = process.argv.slice(2);
if (args.length === 0) {
  console.error("Uso: bun agente_completo.ts <ruta_archivo> [directorio]");
  process.exit(1);
}

revisar(args[0], args[1] || process.cwd())
  .then((resultado) => console.log(JSON.stringify(resultado, null, 2)))
  .catch((err) => {
    console.error("Error:", err.message);
    process.exit(1);
  });

File: go/16-proyecto/agente_completo.go
// Pipeline del agente de revisión de código — equivalente lógico al de Python y TypeScript.

// Cómo ejecutar: make go FILE=go/16-proyecto/agente_completo.go

package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)
var (
	model  = envOr("MODEL", "claude-sonnet-4-6")
	apiURL = envBaseURL()
)

const systemPrompt = `Eres un agente de revisión de código Python.
Analiza el código con cuidado, usa las herramientas disponibles para verificar comportamiento
cuando sea útil, y produce una revisión técnica estructurada.

Cuando tengas suficiente información, emite el resultado final como JSON con este schema exacto:
{
  "hallazgos": [
    {
      "linea": <número o null>,
      "severidad": "<critical|high|medium|low>",
      "tipo": "<bug|estilo|rendimiento|seguridad>",
      "descripcion": "<descripción concisa del hallazgo>",
      "sugerencia": "<cómo corregirlo>"
    }
  ],
  "resumen": "<párrafo de resumen>"
}`

// ---- Tipos de la API de Anthropic ----

type AnthropicRequest struct {
	Model     string      `json:"model"`
	MaxTokens int         `json:"max_tokens"`
	System    string      `json:"system"`
	Tools     []Tool      `json:"tools"`
	Messages  []Message   `json:"messages"`
}

type Tool struct {
	Name        string      `json:"name"`
	Description string      `json:"description"`
	InputSchema interface{} `json:"input_schema"`
}

type Message struct {
	Role    string      `json:"role"`
	Content interface{} `json:"content"`
}

type ContentBlock struct {
	Type  string          `json:"type"`
	Text  string          `json:"text,omitempty"`
	ID    string          `json:"id,omitempty"`
	Name  string          `json:"name,omitempty"`
	Input json.RawMessage `json:"input,omitempty"`
}

type ToolResult struct {
	Type      string `json:"type"`
	ToolUseID string `json:"tool_use_id"`
	Content   string `json:"content"`
}

type AnthropicResponse struct {
	Content    []ContentBlock `json:"content"`
	StopReason string         `json:"stop_reason"`
	Usage      struct {
		InputTokens  int `json:"input_tokens"`
		OutputTokens int `json:"output_tokens"`
	} `json:"usage"`
}

// ---- Herramientas ----

var herramientas = []Tool{
	{
		Name:        "read_file",
		Description: "Lee el contenido de un archivo del proyecto",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"path": map[string]interface{}{"type": "string", "description": "Ruta relativa al directorio del proyecto"},
			},
			"required": []string{"path"},
		},
	},
	{
		Name:        "run_code",
		Description: "Ejecuta un fragmento de código Python en sandbox y devuelve stdout/stderr",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"code":    map[string]interface{}{"type": "string"},
				"timeout": map[string]interface{}{"type": "integer", "default": 10},
			},
			"required": []string{"code"},
		},
	},
	{
		Name:        "search_docs",
		Description: "Busca en la documentación técnica del equipo",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"query": map[string]interface{}{"type": "string"},
			},
			"required": []string{"query"},
		},
	},
	{
		Name:        "write_report",
		Description: "Escribe el informe final de revisión en disco",
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"content":  map[string]interface{}{"type": "string"},
				"filename": map[string]interface{}{"type": "string"},
			},
			"required": []string{"content", "filename"},
		},
	},
}

func ejecutarHerramienta(nombre string, paramsRaw json.RawMessage, proyectoDir string) string {
	var params map[string]interface{}
	json.Unmarshal(paramsRaw, &params)

	switch nombre {
	case "read_file":
		rutaRel, _ := params["path"].(string)
		rutaAbs, err := filepath.Abs(filepath.Join(proyectoDir, rutaRel))
		if err != nil || !strings.HasPrefix(rutaAbs, filepath.Clean(proyectoDir)) {
			return "Error: ruta fuera del directorio del proyecto"
		}
		contenido, err := os.ReadFile(rutaAbs)
		if err != nil {
			return fmt.Sprintf("Error: archivo '%s' no encontrado", rutaRel)
		}
		return string(contenido)

	case "run_code":
		codigo, _ := params["code"].(string)
		timeoutSec := 10
		if t, ok := params["timeout"].(float64); ok {
			timeoutSec = int(t)
		}
		tmpdir, _ := os.MkdirTemp("", "agente-")
		defer os.RemoveAll(tmpdir)
		script := filepath.Join(tmpdir, "script.py")
		os.WriteFile(script, []byte(codigo), 0644)

		cmd := exec.Command("python3", script)
		cmd.Dir = tmpdir
		cmd.Env = []string{"PATH=/usr/local/bin:/usr/bin:/bin", "HOME=" + tmpdir}

		var stdout, stderr bytes.Buffer
		cmd.Stdout = &stdout
		cmd.Stderr = &stderr

		done := make(chan error, 1)
		cmd.Start()
		go func() { done <- cmd.Wait() }()

		select {
		case <-time.After(time.Duration(timeoutSec) * time.Second):
			cmd.Process.Kill()
			return "Error: timeout de ejecución"
		case <-done:
		}

		result := map[string]string{
			"stdout": strings.TrimSpace(stdout.String()),
			"stderr": strings.TrimSpace(stderr.String()),
		}
		if result["stdout"] == "" {
			result["stdout"] = "(vacío)"
		}
		j, _ := json.Marshal(result)
		return string(j)

	case "search_docs":
		query, _ := params["query"].(string)
		return fmt.Sprintf("[Documentación para '%s': ver /docs/ del proyecto]", query)

	case "write_report":
		content, _ := params["content"].(string)
		filename, _ := params["filename"].(string)
		reportsDir := filepath.Join(proyectoDir, "reports")
		os.MkdirAll(reportsDir, 0755)
		ruta := filepath.Join(reportsDir, filename)
		os.WriteFile(ruta, []byte(content), 0644)
		return fmt.Sprintf("Informe escrito en %s", ruta)
	}

	return fmt.Sprintf("Error: herramienta '%s' desconocida", nombre)
}

// ---- Llamada a la API ----

func llamarAnthropic(req AnthropicRequest) (*AnthropicResponse, error) {
	body, _ := json.Marshal(req)
	httpReq, _ := http.NewRequest("POST", apiURL, bytes.NewReader(body))
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("x-api-key", os.Getenv("ANTHROPIC_API_KEY"))
	httpReq.Header.Set("anthropic-version", "2023-06-01")

	resp, err := http.DefaultClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("API error %d: %s", resp.StatusCode, respBody)
	}

	var result AnthropicResponse
	json.Unmarshal(respBody, &result)
	return &result, nil
}

// ---- Loop ReAct ----

func extraerJSON(texto string) (map[string]interface{}, error) {
	for i, c := range texto {
		if c != '{' {
			continue
		}
		dec := json.NewDecoder(strings.NewReader(texto[i:]))
		var result map[string]interface{}
		if err := dec.Decode(&result); err == nil {
			return result, nil
		}
	}
	return nil, fmt.Errorf("no se encontró JSON en output: %s", texto[:min(300, len(texto))])
}

func loopReact(codigo, proyectoDir string) (map[string]interface{}, error) {
	mensajes := []Message{
		{Role: "user", Content: fmt.Sprintf("Revisa este código:\n\n```python\n%s\n```", codigo)},
	}

	const maxPasos = 15

	for paso := 0; paso < maxPasos; paso++ {
		req := AnthropicRequest{
			Model:     model,
			MaxTokens: 4096,
			System:    systemPrompt,
			Tools:     herramientas,
			Messages:  mensajes,
		}

		resp, err := llamarAnthropic(req)
		if err != nil {
			return nil, err
		}

		if resp.StopReason == "end_turn" {
			for _, bloque := range resp.Content {
				if bloque.Type == "text" {
					return extraerJSON(bloque.Text)
				}
			}
			return nil, fmt.Errorf("respuesta sin bloque de texto")
		}

		if resp.StopReason == "tool_use" {
			// Añadir respuesta del asistente
			mensajes = append(mensajes, Message{Role: "assistant", Content: resp.Content})

			// Ejecutar herramientas y recoger resultados
			var resultados []ToolResult
			for _, bloque := range resp.Content {
				if bloque.Type == "tool_use" {
					resultado := ejecutarHerramienta(bloque.Name, bloque.Input, proyectoDir)
					resultados = append(resultados, ToolResult{
						Type:      "tool_result",
						ToolUseID: bloque.ID,
						Content:   resultado,
					})
				}
			}
			mensajes = append(mensajes, Message{Role: "user", Content: resultados})
		}
	}

	return nil, fmt.Errorf("el agente no terminó en %d pasos", maxPasos)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// ---- Memoria episódica ----

type EntradaDB struct {
	Fecha    string                 `json:"fecha"`
	Revision map[string]interface{} `json:"revision"`
}

func cargarDB(dbPath string) map[string]EntradaDB {
	data, err := os.ReadFile(dbPath)
	if err != nil {
		return map[string]EntradaDB{}
	}
	var db map[string]EntradaDB
	json.Unmarshal(data, &db)
	return db
}

func guardarDB(dbPath string, db map[string]EntradaDB) {
	data, _ := json.MarshalIndent(db, "", "  ")
	os.WriteFile(dbPath, data, 0644)
}

func hashCodigo(ruta, codigo string) string {
	h := sha256.Sum256([]byte(ruta + "::" + codigo))
	return fmt.Sprintf("%x", h)
}

// ---- Pipeline principal ----

func revisar(rutaArchivo, proyectoDir string) (map[string]interface{}, error) {
	contenido, err := os.ReadFile(filepath.Join(proyectoDir, rutaArchivo))
	if err != nil {
		return nil, fmt.Errorf("no se pudo leer '%s': %w", rutaArchivo, err)
	}
	codigo := string(contenido)
	hash := hashCodigo(rutaArchivo, codigo)

	dbPath := filepath.Join(proyectoDir, "revisiones.json")
	db := cargarDB(dbPath)

	if entrada, ok := db[hash]; ok {
		fmt.Fprintf(os.Stderr, "[INFO] Revisión cacheada del %s\n", entrada.Fecha)
		entrada.Revision["_cached"] = true
		return entrada.Revision, nil
	}

	revision, err := loopReact(codigo, proyectoDir)
	if err != nil {
		return nil, err
	}

	db[hash] = EntradaDB{
		Fecha:    time.Now().Format(time.RFC3339),
		Revision: revision,
	}
	guardarDB(dbPath, db)

	return revision, nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "Uso: go run agente_completo.go <ruta_archivo> [directorio]")
		os.Exit(1)
	}

	rutaArchivo := os.Args[1]
	proyectoDir := "."
	if len(os.Args) >= 3 {
		proyectoDir = os.Args[2]
	}

	resultado, err := revisar(rutaArchivo, proyectoDir)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
		os.Exit(1)
	}

	output, _ := json.MarshalIndent(resultado, "", "  ")
	fmt.Println(string(output))
}

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

func envBaseURL() string {
	if v := os.Getenv("ANTHROPIC_BASE_URL"); v != "" {
		return v + "/v1/messages"
	}
	return "https://api.anthropic.com/v1/messages"
}
